README.md
#Mozzi  

###sound synthesis library for Arduino  


Tim Barrass  
version 2013-07-13-17:27  

Currently your Arduino can only beep like a microwave oven. Mozzi brings
your Arduino to life by allowing it to produce much more complex and interesting
growls, sweeps and chorusing atmospherics. These sounds can be quickly and easily
constructed from familiar synthesis units like oscillators, delays, filters and
envelopes.

You can use Mozzi to generate algorithmic music for an installation or
performance, or make interactive sonifications of sensors, on a small, modular
and super cheap Arduino, without the need for additional shields, message
passing or external synths.

***

## Features  
-    16384 Hz sample rate, with 8 bit or 14 bit audio output modes.
-    Variable control rate from 64 Hz upwards.
-    Useful basic audio toolkit: oscillators, samples, lines, envelopes, scheduling, filtering.
-    Fast ADC and other cpu-efficient code utilities to help keep audio running smoothly.
-    Example sketches for easy modification.
-    Readymade wavetables and a script to convert your own soundfiles for Mozzi.  
-	 Mozzi is designed to be easy to use, open source and extendable.

***

## Installation  
Download Mozzi from the top of this page and unzip it. It will probably have a
name like "sensorium-Mozzi-2bee818". Rename the unzipped folder "Mozzi". Then,
following the instructions from the [Arduino libraries guide](http://arduino.cc/en/Guide/Libraries):  

In the Arduino IDE, navigate to __Sketch > Import Library__. At the top of the drop
down list, select the option to __Add Library__. Return to the __Sketch > Import Library__ menu. 
You should now see the library at the bottom of the drop-down
menu. It is ready to be used in your sketch.

***

## Quick Start  
To hear Mozzi, connect a 3.5mm audio jack with the centre wire to the PWM output
on Digital Pin 9\* on Arduino, and the ground to the Ground on the Arduino.
Use this as a line out which you can plug into your computer and listen to with
a sound program like [Audacity](http://audacity.sourceforge.net/). 
Try some examples from the __File > Examples > Mozzi__ menu.  
* See the README for audio output pins on different boards. 

Below is a list of the Digital Pins used by Mozzi for STANDARD mode PWM audio out on different boards.
Those which have been tested and reported to work have an x.
Feedback about others is welcome.

x	 9	Arduino Uno  
x	 9	Arduino Duemilanove  
x	 9	Arduino Nano  
x	 9	Arduino Pro Mini  
x	 9	Arduino Leonardo
x	11	Arduino Mega  
x	11  Freetronics EtherMega  
x	 9  Ardweeny  
x	 9  Boarduino
..14	Teensy  
x	B5  Teensy2  
x	B5(25) Teensy2++  
..13	Sanguino  

For details about HIFI mode, read the [Mozzi core module documentation](http://sensorium.github.com/Mozzi/doc/html/group__core.html#gae99eb43cb29bb03d862ae829999916c4/).  

***

## Using Mozzi  
Here's a template for an empty Mozzi sketch:  

```
#include <MozziGuts.h>   // at the top of your sketch

void setup() {
	startMozzi();
}

void updateControl(){
	// your control code
}

int updateAudio(){
	// your audio code which returns an int between -244 and 243
}

void loop() {
	audioHook();
}
```

There's a detailed example explaining the different parts [here](http://sensorium.github.com/Mozzi/learn/a-simple-sketch/).

***

## Documentation

There's documentation in the doc folder in the Mozzi download and [online](http://sensorium.github.com/Mozzi/doc/html/index.html).  
There is practical help on the [learn](http://sensorium.github.com/Mozzi/learn/) page on the Mozzi site.  
Start or look up a topic on the [users forum](https://groups.google.com/forum/#!forum/mozzi-users/).  
Also, feel free to submit any issues on the [GitHub Mozzi site](https://github.com/sensorium/Mozzi/issues/).  
Look for code and usage changes [here](extras/NEWS.txt).  

***

## Caveats

While Mozzi is running, the Arduino time functions `millis()`, `micros()`, `delay()`, and
`delayMicroseconds()` are disabled. Instead, Mozzi provides `mozziMicros()` for timing, with 61us resolution (in `STANDARD` mode), and
`EventDelay()` for scheduling.  Also, there's `pauseMozzi()` and `unpauseMozzi()` if the Arduino timers are required for other things.  

***
If you enjoy using Mozzi for a project, or have extended it, we would be
pleased to hear about it and provide support wherever possible. Contribute
suggestions, improvements and bug fixes to the Mozzi wiki on Github, or
Fork it to contribute directly to future developments.

Mozzi is a development of research into Mobile Sonification in the
[SweatSonics](http://stephenbarrass.wordpress.com/tag/sweatsonics/) project. 

***

##Contributions / Included Dependencies  
Modified versions of the following libraries are included in the Mozzi download:  

[TimerOne library](http://www.pjrc.com/teensy/td_libs_TimerOne.html)  
[FrequencyTimer2 library](http://www.pjrc.com/teensy/td_libs_FrequencyTimer2.html) - now a [fork with support for ATmega32u4 processors](https://github.com/sensorium/FrequencyTimer2/)   

Mozzi has also drawn on and been influenced by (among many others):  

[xorshift] (http://www.jstatsoft.org/v08/i14/xorshift.pdf) random number generator, George Marsaglia, (2003)  
ead~.c puredata external (creb library) Copyright (c) 2000-2003 by Tom Schouten (GPL2)  
[AF_precision_synthesis](http://adrianfreed.com/content/arduino-sketch-high-frequency-precision-sine-wave-tone-sound-synthesis)
by Adrian Freed, 2009  
[Resonant filter](http://www.musicdsp.org/archive.php?classid=3#259) posted to musicdsp.org by Paul Kellett,
and fixed point version of the filter on [dave's blog of art and programming] (http://www.pawfal.org/dave/blog/2011/09/)  
State Variable filter pseudocode at [musicdsp.org] (http://www.musicdsp.org/showone.php?id=23 and http://www.musicdsp.org/showone.php?id=142)  
Various examples from [Pure Data](http://puredata.info/) by Miller Puckette  
[Practical synthesis tutorials](http://www.obiwannabe.co.uk/) by Andy Farnell  


NEWS.txt
Mozzi sound synthesis library for Arduino

NEWS
get the latest version from http://sensorium.github.io/Mozzi/


version_2013-07-07-17:51
- had #define AUDIO_MODE HIFI as default - changed back to #define AUDIO_MODE STANDARD,
	it would have made most examples sound bad without any warning.  sorry
- new versioning: a timestamp


_Version 0.01.2e_
- bugfixes.  Should have tested properly!  
- Fixed ADSR compile problem due to mistakes while editing docs.
- Fixed Table_Resolution compile prob due to sin table error
- Added AutoMap.h, for mapping unpredictable sensor input ranges to a set output range 
	- demos of AutoMap in examples/sensors/Knob_LDR_x2_WavePacket and a couple of others in sensors/


Version 0.01.2d
- mozzi_analog.cpp - fixed show-stopping typo (!= instead of |=)in adcStartConversion() and adcStartConversion(pin).
- added WavePacketSample.h, a WavePacket which allows a custom table for audio source (limited at the moment to 8192 samples)
	- added samples/raven_arh_int8.h to test WavePacketSample with.
- lots of minor documentation fixes, new layout.
- re-arranged website with some docs more easy to find.


Version 0.01.2c
- removed Ead.cpp, condensed to Ead.h
- mozzi_analog.h changed all xorshift96()>>16's to xorshift96() & 0xFFFF and all /256's to >>8
- added ControlDelay(), the same as AudioDelay() but for control values.
- added AutoRange(), for keeping a running calculation of sensor input range (should also have a variable timespan/fading memory).
- tweaked twi_nonblock.cpp and twi_nonblock.h to get rid of some compile warnings about return values.
	Madetwi_initiateReadFrom() and twi_initiateWriteFrom() return their error states 
	and made twi_continueReadFrom() and twi_continueWriteFrom() return void.
- numbered the Examples category folders to give more sense of organisation and progression
- added Examples > Mozzi > _04_analog_input > Audio_Input, a minimal test of Mozzi audio input.
- added Examples  > Mozzi > _02_control > Control_Echo_Theremin, demonstrating ControlDelay()
- repaired hack for Teensy 2 in mozzi_analog.h so it compiles again
- moved default CONTROL_RATE definition to MozziGuts.h, it doesn't need to be in 
    a config file if you can change it in a sketch.
- changed sin1024_int8.h table to use full range signed 8 bit values - it was scaled down slightly for a forgotten project
- removed samples/TJBAnalogDisk and samples/thumbpiano, these will be available as a separate "Mozzi_Extras" download.
- removed samples/Bamboo1 and changed samples/Samples sketch to use samples/bamboo

    
Version 0.01.2b
- fixed some docs
- added thumbpiano samples


Version 0.01.2a
- Added Examples>Mozzi>communication>Sinewave_HIFI_PWM_leds, to show 
	non-blocking pwm for any pins without disprupting audio.
- Added Examples>Mozzi>sensors directory with several basic examples 
	for modifiying sound with light dependent resistors, potentiometers, piezos and switches.
	These go with a tutorial on the main Mozzi site http://sensorium.github.io/Mozzi/Mozzi_Introductory_Tutorial.pdf
	Updated from previous unannounced inclusion which had lots of errors...


Version 0.01.1zzzz
- fixed indent error in all python scripts (error introduced in last release)
- removed a wierd file link which stopped installation working
- added some samples from a collection of analog bass samples by T.J.Baker,
	from Amiga aminet archive sample disk TJBAnalogDisk.lha.  
	Samples are in Mozzi/samples/TJBAnalogDisk.  
	There are wav files to audition and converted versions for Mozzi.  
	More of the like to come...


Version 0.01.1zzz
- Modified Flexitimer2.cpp so ATMEGA32U4 processor based boards work again (Leonardo,Teensy2.0).
- Included twi_nonblock.h and twi_nonblock.cpp, non-blocking two wire I2C code by Marije Baalman, 
	with Examples>Mozzi>Communication>TwoWire_Read_ADXL345.  Not documented yet.
- Sample.h 
	- added isPlaying(), true if the sample is playing
	- fixed rangeWholeSample() so startpos = 0
- changed all instances of #include Arduino.h to conditional include WProgram.h for older versions of the IDE..not tested
- Added Examples>Mozzi>communication>Mozzi_Processing_Serial, 
	demonstrates the sending data from the computer to the Arduino board.
- Added Examples>Mozzi>control>RCpoll, demonstrates reading a resistive sensor on digital pin without blocking audio.



Version 0.01.1zz
- Added ReverbTank, a reverb which sounds like the inside of a tin can, and an example ReverbTank_STANDARD
- AudioDelay 
	- added template parameter for signal type, to allow ints as well as the default char type.
	- next() now returns only delayed signal, so mixing can be flexible and is done in the sketch
- AudioDelayFeedback 
	- repaired a few bugs. (I hope)
	- added next(char input) which was missing for LINEAR interpolated delays
	
	
	
Version 0.01.1z
- Phasor - changed setFreq(unsigned int) to setFreq(int), old sketches will need updating
- Reorganised examples into topic subdirectories and removed numbers
- MozziGuts.cpp 
	- Added pauseMozzi() - Stops audio and control interrupts and restores the timers to the values they had before Mozzi was started.
	This will re-enable the Arduino time functions delay() and delayMicroseconds() while paused.
	- Added unPauseMozzi(), allowing Mozzi to continue again after being paused.
	- added mozziMicros(), a replacement for Arduino micros().
- mozzi_config.h - AUDIO_RATE is now a configuration option which can be defined as 16384 (default) or 32768.
	16384 obviously allows more time for each sample to be calculated, but the higher rate may 
	be useful in some situations (experimental).
- AudioDelayFeedback.h 
	- added optional template parameters for LINEAR (default) or ALLPASS interpolation.
	- added examples/delays/AudioFeedbackDelayAllpass example.
- Added WavePacket.h, a synthesis unit based on Miller Puckette's wavepacket example in Pure Data.
	- added examples/synthesis/WavePacket_Double and WavePacket_Single examples
- mozzi_fixmath.h 

Oscil.h
/*
 * Oscil.h
 *
 * Oscil.h owes much to AF_precision_synthesis.pde, 2009, Adrian Freed.
 *
 * Copyright 2012 Tim Barrass, 2009 Adrian Freed.
 *
 * This file is part of Mozzi.
 *
 * Mozzi is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Mozzi is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Mozzi.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

#ifndef OSCIL_H_
#define OSCIL_H_

#if ARDUINO >= 100
 #include "Arduino.h"
#else
 #include "WProgram.h"
#endif
#include "MozziGuts.h"
#include "mozzi_fixmath.h"
#include <util/atomic.h>


#ifdef OSCIL_DITHER_PHASE
#include "mozzi_rand.h"
#endif

// fractional bits for oscillator index precision
#define OSCIL_F_BITS 16
#define OSCIL_F_BITS_AS_MULTIPLIER 65536

// phmod_proportion is an 1n15 fixed-point number only using
// the fractional part and the sign bit
#define OSCIL_PHMOD_BITS 16



/** 
Oscil plays a wavetable, cycling through the table to generate an audio or
control signal. The frequency of the signal can be set or changed with
setFreq(), and the output of an Oscil can be produced with next() for a simple
cycling oscillator, or atIndex() for a particular sample in the table.
@tparam NUM_TABLE_CELLS This is defined in the table ".h" file the Oscil will be
using. It's important that it's a power of 2, and either a literal number (eg. "8192") or a
defined macro, rather than a const or int, for the Oscil to run fast enough.
@tparam UPDATE_RATE This will be AUDIO_RATE if the Oscil is updated in
updateAudio(), or CONTROL_RATE if it's updated each time updateControl() is
called. It could also be a fraction of CONTROL_RATE if you are doing some kind
of cyclic updating in updateControl(), for example, to spread out the processor load.
@todo Use conditional compilation to optimise setFreq() variations for different table
sizes.
@note If you #define OSCIL_DITHER_PHASE before you #include <Oscil.h>,
the phase increments will be dithered, which reduces spurious frequency spurs
in the audio output, at the cost of some extra processing and memory.
@section char2mozzi
Converting soundfiles for Mozzi
There is a python script called char2mozzi.py in the Mozzi/python folder.
The usage is:
python char2mozzi.py infilename outfilename tablename samplerate
*/
//template <unsigned int NUM_TABLE_CELLS, unsigned int UPDATE_RATE, bool DITHER_PHASE=false>
template <unsigned int NUM_TABLE_CELLS, unsigned int UPDATE_RATE>
class Oscil
{


public:
	/** Constructor.
	@param TABLE_NAME the name of the array the Oscil will be using. This
	can be found in the table ".h" file if you are using a table made for
	Mozzi by the char2mozzi.py python script in Mozzi's python
	folder.*/
	Oscil(const char * TABLE_NAME):table(TABLE_NAME)
	{}


	/** Constructor.
	Declare an Oscil with template TABLE_NUM_CELLS and UPDATE_RATE
	parameters, without specifying a particular wave table for it to play.
	The table can be set or changed on the fly with setTable(). Any tables
	used by the Oscil must be the same size.
	*/
	Oscil()
	{}


	/** Updates the phase according to the current frequency
	and returns the sample at the new phase position.
	@return the next sample.
	*/
	inline
	char next()
	{
		incrementPhase();
		return readTable();
	}


	/** Change the sound table which will be played by the Oscil.
	@param TABLE_NAME is the name of the array in the table ".h" file you're using.
	*/
	void setTable(const char * TABLE_NAME)
	{
		table = TABLE_NAME;
	}


	/** Set the phase of the Oscil.  This does the same thing as Sample::start(offset).  Just different ways of thinking about oscillators and samples.
	@param phase a position in the wavetable.
	@todo Test commenting out ATOMIC_BLOCK in setPhase(), setFreq(), etc.
	*/
	// This could be called in the control interrupt, so phase_fractional should really be volatile,
	// but that could limit optimisation.  Since phase_fractional gets changed often in updateAudio()
	// (in loop()), it's probably worth keeping it nonvolatile until it causes problems
	void setPhase(unsigned int phase)
	{
		ATOMIC_BLOCK(ATOMIC_RESTORESTATE)
		{
			phase_fractional = (unsigned long)phase << OSCIL_F_BITS;
		}
	}

	/** Set the phase of the Oscil.  Might be useful with getPhaseFractional().
	@param phase a position in the wavetable.
	@todo Test commenting out ATOMIC_BLOCK in setPhase(), setFreq(), etc.
	*/
	// This could be called in the control interrupt, so phase_fractional should really be volatile,
	// but that could limit optimisation.  Since phase_fractional gets changed often in updateAudio()
	// (in loop()), it's probably worth keeping it nonvolatile until it causes problems
	void setPhaseFractional(unsigned long phase)
	{
		ATOMIC_BLOCK(ATOMIC_RESTORESTATE)
		{
			phase_fractional = phase;
		}
	}


	/** Get the phase of the Oscil in fractional format.
	@return position in the wavetable, shifted left by OSCIL_F_BITS (which is 16 when this was written).
	*/
	unsigned long getPhaseFractional()
	{
		ATOMIC_BLOCK(ATOMIC_RESTORESTATE)
		{
			return phase_fractional;
		}
	}



	/** Returns the next sample given a phase modulation value.
	@param phmod_proportion a phase modulation value given as a proportion of the wave. The
	phmod_proportion parameter is a Q15n16 fixed-point number where the fractional
	n16 part represents -1 to 1, modulating the phase by one whole table length in
	each direction.
	@return a sample from the table.
	@todo does the phmod_proportion param need to be Q15n16, why not Q0n15 and cast to long for multiply?
	*/
	// PM: cos((angle += incr) + change)
	// FM: cos(angle += (incr + change))
	// The ratio of deviation to modulation frequency is called the "index of modulation". ( I = d / Fm )
	inline
	char phMod(Q15n16 phmod_proportion)
	{
		incrementPhase();
		return (char)pgm_read_byte_near(table + (((phase_fractional+(phmod_proportion * NUM_TABLE_CELLS))>>OSCIL_F_BITS) & (NUM_TABLE_CELLS - 1)));
	}


	/** Set the oscillator frequency with an unsigned int. This is faster than using a
	float, so it's useful when processor time is tight, but it can be tricky with
	low and high frequencies, depending on the size of the wavetable being used. If
	you're not getting the results you expect, try explicitly using a float, or try
	setFreq_Q24n8() or or setFreq_Q16n16().
	@param frequency to play the wave table.
	*/
	inline
	void setFreq (int frequency) {
		ATOMIC_BLOCK(ATOMIC_RESTORESTATE)
		{
			phase_increment_fractional = ((((unsigned long)NUM_TABLE_CELLS<<ADJUST_FOR_NUM_TABLE_CELLS)*frequency)/UPDATE_RATE) << (OSCIL_F_BITS - ADJUST_FOR_NUM_TABLE_CELLS);
		}
	}


	/** Set the oscillator frequency with a float. Using a float is the most reliable
	way to set frequencies, -Might- be slower than using an int but you need either
	this, setFreq_Q24n8() or setFreq_Q16n16() for fractional frequencies.
	@param frequency to play the wave table.
	 */
	inline
	void setFreq(float frequency)
	{ // 1 us - using float doesn't seem to incur measurable overhead with the oscilloscope
		ATOMIC_BLOCK(ATOMIC_RESTORESTATE)
		{
			phase_increment_fractional = (unsigned long)((((float)NUM_TABLE_CELLS * frequency)/UPDATE_RATE) * OSCIL_F_BITS_AS_MULTIPLIER);
		}
	}


	/** Set the frequency using Q24n8 fixed-point number format.
	This might be faster than the float version for setting low frequencies such as
	1.5 Hz, or other values which may not work well with your table size. A Q24n8
	representation of 1.5 is 384 (ie. 1.5 * 256). Can't be used with UPDATE_RATE
	less than 64 Hz.
	@param frequency in Q24n8 fixed-point number format.
	*/
	inline
	void setFreq_Q24n8(Q24n8 frequency)
	{
		ATOMIC_BLOCK(ATOMIC_RESTORESTATE)
		{
			//phase_increment_fractional = (frequency* (NUM_TABLE_CELLS>>3)/(UPDATE_RATE>>6)) << (F_BITS-(8-3+6));
			phase_increment_fractional = (((((unsigned long)NUM_TABLE_CELLS<<ADJUST_FOR_NUM_TABLE_CELLS)>>3)*frequency)/(UPDATE_RATE>>6))
			                             << (OSCIL_F_BITS - ADJUST_FOR_NUM_TABLE_CELLS - (8-3+6));
		}
	}


	/** Set the frequency using Q16n16 fixed-point number format. This is useful in
	combination with Q16n16_mtof(), a fast alternative to mtof(), using Q16n16
	fixed-point format instead of floats.
	@note This should work OK with tables 2048 cells or smaller and
	frequencies up to 4096 Hz.  Can't be used with UPDATE_RATE less than 64 Hz.
	@param frequency in Q16n16 fixed-point number format.
	*/
	inline
	void setFreq_Q16n16(Q16n16 frequency)
	{
		ATOMIC_BLOCK(ATOMIC_RESTORESTATE)
		{
			//phase_increment_fractional = ((frequency * (NUM_TABLE_CELLS>>7))/(UPDATE_RATE>>6)) << (F_BITS-16+1);
			phase_increment_fractional = (((((unsigned long)NUM_TABLE_CELLS<<ADJUST_FOR_NUM_TABLE_CELLS)>>7)*frequency)/(UPDATE_RATE>>6))
			                             << (OSCIL_F_BITS - ADJUST_FOR_NUM_TABLE_CELLS - 16 + 1);

		}
	}
/*
	inline
	void setFreqMidi(char note_num) {
		setFreq_Q16n16(mtof(note_num));
	}
*/
	/**  Returns the sample at the given table index.
	@param index between 0 and the table size.The
	index rolls back around to 0 if it's larger than the table size.
	@return the sample at the given table index.
	*/
	inline
	char atIndex(unsigned int index)
	{
		return (char)pgm_read_byte_near(table + (index & (NUM_TABLE_CELLS - 1)));
	}


	/** phaseIncFromFreq() and setPhaseInc() are for saving processor time when sliding
	between frequencies. Instead of recalculating the phase increment for each
	frequency in between, you can just calculate the phase increment for each end
	frequency with phaseIncFromFreq(), then use a Line to interpolate on the fly and
	use setPhaseInc() to set the phase increment at each step. (Note: I should
	really profile this with the oscilloscope to see if it's worth the extra
	confusion!)
	@param frequency for which you want to calculate a phase increment value.
	@return the phase increment value which will produce a given frequency.
	*/
	inline
	//const
	unsigned long phaseIncFromFreq(int frequency)
	{
		return (((unsigned long)frequency * NUM_TABLE_CELLS)/UPDATE_RATE) << OSCIL_F_BITS;
	}


	/** Set a specific phase increment.  See phaseIncFromFreq().
	@param phaseinc_fractional a phase increment value as calculated by phaseIncFromFreq().
	 */
	inline
	void setPhaseInc(unsigned long phaseinc_fractional)
	{
		ATOMIC_BLOCK(ATOMIC_RESTORESTATE)
		{
			phase_increment_fractional = phaseinc_fractional;
		}
	}



private:


	/** Used for shift arithmetic in setFreq() and its variations.
	*/
static const unsigned char ADJUST_FOR_NUM_TABLE_CELLS = (NUM_TABLE_CELLS<2048) ? 8 : 0;


	/** Increments the phase of the oscillator without returning a sample.
	 */
	inline
	void incrementPhase()
	{
		//phase_fractional += (phase_increment_fractional | 1); // odd phase incr, attempt to reduce frequency spurs in output
		phase_fractional += phase_increment_fractional;
	}


	/** Returns the current sample.
	 */
	inline
	char readTable()
	{
#ifdef OSCIL_DITHER_PHASE
		return (char)pgm_read_byte_near(table + (((phase_fractional + ((int)(xorshift96()>>16))) >> OSCIL_F_BITS) & (NUM_TABLE_CELLS - 1)));
#else
		return (char)pgm_read_byte_near(table + ((phase_fractional >> OSCIL_F_BITS) & (NUM_TABLE_CELLS - 1)));
		//return (char)pgm_read_byte_near(table + (((phase_fractional >> OSCIL_F_BITS) | 1 ) & (NUM_TABLE_CELLS - 1))); odd phase, attempt to reduce frequency spurs in output
#endif
	}


	unsigned long phase_fractional;
	volatile unsigned long phase_increment_fractional; // volatile with atomic access because it can
	// be set in the updateControl() interrupt and
	// used in updateAudio(), which is outside the
	// interrupt.
	const char * table;

};


#endif /* OSCIL_H_ */
